### PR TITLE
PXP-6106 fix: auth cmd authz section

### PR DESCRIPTION
**Pull request by mfshao**
Jira Ticket: [PXP-6106](https://ctds-planx.atlassian.net/browse/PXP-6106)

`auth` command will checks for `authz` section (Arborist premissions) first, and fallback to check for `project_access` section (Fence premissions) if `authz` section is empty or missing

```
➜  bin ./gen3-client auth --profile=mshao_qa-mickey
2021/03/02 11:21:17 
You have access to the following resource(s) at https://qa-mickey.planx-pla.net:
2021/03/02 11:21:17 /dashboard [
  {
    "method": "access",
    "service": "dashboard"
  }
]
2021/03/02 11:21:17 /data_file [
  {
    "method": "file_upload",
    "service": "fence"
  }
]
2021/03/02 11:21:17 /gen3/programs/DEV [
  {
    "method": "create",
    "service": "*"
  },
  {
    "method": "delete",
    "service": "*"
  },
  {
    "method": "read",
    "service": "*"
  },
  {
    "method": "read-storage",
    "service": "*"
  },
  {
    "method": "update",
    "service": "*"
  }
]
...
```

### Improvements
- `auth` command will checks for `authz` section (Arborist premissions) first, and fallback to check for `project_access` section (Fence premissions) if `authz` section is empty or missing

### CODE REVIEW

File: tests/functions_test.go [hunk 1]
```diff
@@ -178,17 +179,17 @@ func TestCheckPrivilegesGrantedAccess(t *testing.T) {
...
 
-	grantedAccessJson := "{\"project_access\": " +
-		"{\"test_project\": {" +
-		"\"0\": \"read\"," +
-		"\"1\": \"create\"," +
-		"\"2\": \"read-storage\"," +
-		"\"3\": \"update\"," +
-		"\"4\": \"delete\"}" +
-		"}}"
+	grantedAccessJSON := "{\"project_access\": " +
```
Comment (williamhaley) [inline tests/functions_test.go:182]:
```suggestion
	grantedAccessJSON := `{
		"project_access": {
			"test_project": ["read","create","read-storage","update","delete"]
		}
	}`
```

Not really related to this PR, but template strings might make this JSON a little cleaner 🧹

File: tests/functions_test.go [hunk 2]
```diff
@@ -213,3 +214,63 @@ func TestCheckPrivilegesGrantedAccess(t *testing.T) {
...
+func TestCheckPrivilegesGrantedAccessAuthz(t *testing.T) {
+
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+
+	mockConfig := mocks.NewMockConfigureInterface(mockCtrl)
+	mockRequest := mocks.NewMockRequestInterface(mockCtrl)
+	testFunction := &jwt.Functions{Config: mockConfig, Request: mockRequest}
+
+	cred := jwt.Credential{KeyId: "", APIKey: "fake_api_key", AccessKey: "non_exprired_token", APIEndpoint: "http://www.test.com"}
```
Comment (ZakirG) [inline tests/functions_test.go:228]: typo - "exprired"

File: tests/functions_test.go [hunk 3]
```diff
@@ -213,3 +214,63 @@ func TestCheckPrivilegesGrantedAccess(t *testing.T) {
...
+		"\"delete\"]}" +
+		"}"
+
+	mockedResp := &http.Response{
+		Body:       ioutil.NopCloser(bytes.NewBufferString(grantedAccessJSON)),
+		StatusCode: 200,
+	}
+
+	mockConfig.EXPECT().ParseConfig("default").Return(cred).Times(1)
+	mockRequest.EXPECT().MakeARequest("GET", "http://www.test.com/user/user", "non_exprired_token", "", gomock.Any(), gomock.Any()).Return(mockedResp, nil).Times(1)
```
Comment (ZakirG) [inline tests/functions_test.go:254]: typo - "exprired"

File: tests/functions_test.go [hunk 4]
```diff
@@ -213,3 +214,63 @@ func TestCheckPrivilegesGrantedAccess(t *testing.T) {
...
+		{"method": "read-storage", "service": "*"},
+		{"method": "update", "service": "*"},
+		{"method": "upload", "service": "*"},
+	}
+
+	if err != nil {
+		t.Errorf("Expected no errors, received an error \"%v\"", err)
+		// don't use DeepEqual since expectedAccess is []interface {} and receivedAccess is []map[string]interface {}, just check for contents
+	} else if fmt.Sprint(expectedAccess) != fmt.Sprint(receivedAccess) {
+		t.Errorf(`Expected user access and received user access are note the same.
```
Comment (ZakirG) [inline tests/functions_test.go:272]: typo - "note"

File: gen3-client/g3cmd/auth.go [hunk 5]
```diff
@@ -27,33 +28,42 @@ func init() {
...
+							for _, permission := range permissions {
+								access = append(access, permission.(string))
+							}
+							sort.Strings(access)
+							log.Printf("%s %s\n", project, access)
+						} else {
+							// Premissions from Arborist already sorted, just pretty print them
+							marshalledPermissions, err := json.MarshalIndent(permissions, "", "  ")
+							if err != nil {
+								log.Printf("%s (error occurred when marshalling premissions): %s\n", project, err)
```
Comment (williamhaley) [inline gen3-client/g3cmd/auth.go:63]:
```suggestion
								log.Printf("%s (error occurred when marshalling permissions): %s\n", project, err)
```

File: gen3-client/jwt/functions.go [hunk 6]
```diff
@@ -296,12 +296,17 @@ func (f *Functions) CheckPrivileges(profile string, configFileType string) (stri
 		return "", nil, errors.New("Error occurred when unmarshalling response: " + err.Error())
 	}
 
-	projectAccess, ok := data["project_access"].(map[string]interface{})
-	if !ok {
-		return "", nil, errors.New("Not possible to read user access privileges")
+	resourceAccess, ok := data["authz"].(map[string]interface{})
+
+	// If the `authz` section (Arborist permissions) is empty or missing, try get `project_access` section (Fence permissions)
+	if len(resourceAccess) == 0 || !ok {
```
Comment (mfshao) [inline gen3-client/jwt/functions.go:302]: Q for @paulineribeyre maybe: is this fallbeck check vaild? could there be user with empty `authz` section (which is intended), but no `project_access` section?

Comment (paulineribeyre) [reply gen3-client/jwt/functions.go:302]: looks fine to me, `project_access` is always included

**Review by ZakirG — APPROVED** [review]

**Pull request merged**